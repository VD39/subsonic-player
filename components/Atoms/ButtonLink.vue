<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: Component | Icon;
    iconColor?: string;
    iconPosition?: IconPosition;
    iconSize?: IconSize;
    iconWeight?: IconWeight;
    is?: ButtonLink;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconColor: undefined,
    iconPosition: 'left',
    iconSize: 'medium',
    iconWeight: 'regular',
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      'centerItems',
      $style.buttonLink,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        'centerAll fullWidth': fullWidth,
      },
    ]"
    :disabled
    draggable="false"
    :type
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="iconComponent"
      aria-hidden="true"
      :class="$style.icon"
      :color="iconColor"
      :size="ICON_SIZE[iconSize]"
      :weight="iconWeight"
    />

    <span
      ref="text"
      :class="[
        'clamp',
        $style.text,
        {
          visuallyHidden: !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.buttonLink {
  position: relative;
  gap: var(--default-space);
  padding: var(--default-space);
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  flex-shrink: 0;

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
}
</style>
