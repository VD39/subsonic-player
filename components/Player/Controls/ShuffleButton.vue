<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';

const { shuffle, toggleShuffle } = useAudioPlayer();

const buttonProps = computed<ShuffleButtonProps>(() => ({
  iconColor: shuffle.value ? 'var(--theme-color)' : 'currentColor',
  iconWeight: shuffle.value ? 'fill' : 'duotone',
  title: `Switch shuffle ${shuffle.value ? 'on' : 'off'}`,
}));
</script>

<template>
  <IconButton
    icon="PhShuffle"
    :icon-color="buttonProps.iconColor"
    :icon-weight="buttonProps.iconWeight"
    :title="buttonProps.title"
    @click="toggleShuffle"
  >
    {{ buttonProps.title }}
  </IconButton>
</template>
