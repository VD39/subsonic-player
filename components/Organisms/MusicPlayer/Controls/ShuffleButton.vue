<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

const { shuffle, toggleShuffle } = useAudioPlayer();

const buttonProps = computed<ButtonProps>(() => ({
  iconColor: shuffle.value ? 'var(--theme-color)' : 'currentColor',
  iconWeight: shuffle.value ? 'fill' : 'duotone',
  title: `Switch shuffle ${shuffle.value ? 'on' : 'off'}`,
}));
</script>

<template>
  <ButtonLink
    :icon="ICONS.shuffle"
    :icon-color="buttonProps.iconColor"
    :icon-weight="buttonProps.iconWeight"
    :title="buttonProps.title"
    @click="toggleShuffle"
  >
    {{ buttonProps.title }}
  </ButtonLink>
</template>
