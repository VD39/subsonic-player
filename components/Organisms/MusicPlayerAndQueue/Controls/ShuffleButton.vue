<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';

const { shuffle, toggleShuffle } = useAudioPlayer();

const buttonProps = computed<ButtonProps>(() => ({
  iconColor: shuffle.value ? 'var(--theme-color)' : 'currentColor',
  iconWeight: shuffle.value ? 'fill' : 'regular',
  text: `Switch shuffle ${shuffle.value ? 'on' : 'off'}`,
}));
</script>

<template>
  <ButtonLink
    :icon="ICONS.shuffle"
    :iconColor="buttonProps.iconColor"
    :iconWeight="buttonProps.iconWeight"
    :title="buttonProps.text"
    @click="toggleShuffle"
  >
    {{ buttonProps.text }}
  </ButtonLink>
</template>
