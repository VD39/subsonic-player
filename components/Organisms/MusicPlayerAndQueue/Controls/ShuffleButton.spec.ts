import type { VueWrapper } from '@vue/test-utils';

import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import { useAudioPlayerMock } from '@/test/useAudioPlayerMock';
import { mount } from '@vue/test-utils';

import ShuffleButton from './ShuffleButton.vue';

const { shuffleMock, toggleShuffleMock } = useAudioPlayerMock();

function factory(props = {}) {
  return mount(ShuffleButton, {
    props: {
      ...props,
    },
  });
}

const buttonProps = {
  false: {
    color: 'currentColor',
    text: 'Switch shuffle off',
    weight: 'regular',
  },
  true: {
    color: 'var(--theme-color)',
    text: 'Switch shuffle on',
    weight: 'fill',
  },
};

describe('ShuffleButton', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  describe.each([
    [false, buttonProps.false],
    [true, buttonProps.true],
  ])('when shuffle is set to %s', (shuffle, buttonProps) => {
    beforeEach(() => {
      shuffleMock.value = shuffle;
    });

    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('sets the correct color on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).props('iconColor')).toBe(
        buttonProps.color,
      );
    });

    it('sets the correct weight on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).props('iconWeight')).toBe(
        buttonProps.weight,
      );
    });

    it('sets the correct title attribute on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).attributes('title')).toBe(
        buttonProps.text,
      );
    });

    it('sets the correct slot data on the ButtonLink component', () => {
      expect(wrapper.findComponent(ButtonLink).text()).toContain(
        buttonProps.text,
      );
    });
  });

  describe('when the ButtonLink component is clicked', () => {
    beforeEach(() => {
      wrapper.findComponent(ButtonLink).vm.$emit('click');
    });

    it('calls the toggleShuffle function', () => {
      expect(toggleShuffleMock).toHaveBeenCalled();
    });
  });
});
