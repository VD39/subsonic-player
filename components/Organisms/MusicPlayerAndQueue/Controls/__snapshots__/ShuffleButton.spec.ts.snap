// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShuffleButton > when shuffle is set to false > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" draggable="false" type="button" title="Switch shuffle off">
  <phshuffle aria-hidden="true" class="icon" color="currentColor" size="24" weight="regular"></phshuffle><span class="clamp text visuallyHidden">Switch shuffle off</span>
</button>"
`;

exports[`ShuffleButton > when shuffle is set to true > matches the snapshot 1`] = `
"<button class="centerItems buttonLink" draggable="false" type="button" title="Switch shuffle on">
  <phshuffle aria-hidden="true" class="icon" color="var(--theme-color)" size="24" weight="fill"></phshuffle><span class="clamp text visuallyHidden">Switch shuffle on</span>
</button>"
`;
